svm/src/transaction_processor.rs

```diff
@@ -484,32 +469,65 @@ impl<FG: ForkGraph> TransactionBatchProcessor<FG> {
         let fee_payer_index = 0;
         validate_fee_payer(
             fee_payer_address,
-            &mut fee_payer_account,
+            &mut loaded_fee_payer.account,
             fee_payer_index,
             error_counters,
             rent_collector,
             fee_details.total_fee(),
         )?;
 
+        // If the nonce has been used in this batch already, we must drop the transaction
+        // This is the same as if it was used in different batches in the same slot
+        // If the nonce account was closed in the batch, we error as if the blockhash didn't validate
+        // We must vaidate the account in case it was reopened, either as a normal system account, or a fake nonce account
+        if let Some(ref advanced_nonce_info) = advanced_nonce {
+            let nonces_are_equal = account_loader
+                .load_account(
+                    advanced_nonce_info.address(),
+                    AccountUsagePattern::WritableInstruction,
+                )
+                .and_then(|loaded_nonce| {
+                    let current_nonce_account = &loaded_nonce.account;
+                    system_program::check_id(current_nonce_account.owner()).then_some(())?;
+                    StateMut::<NonceVersions>::state(current_nonce_account).ok()
+                })
+                .and_then(
+                    |current_nonce_versions| match current_nonce_versions.state() {
+                        NonceState::Initialized(ref current_nonce_data) => {
+                            Some(&current_nonce_data.durable_nonce == durable_nonce)
+                        }
+                        _ => None,
+                    },
+                );
```

Comment:
flagging this for special attention because it is quite tricky. i wrote more integration tests for this one block than any other specific thing